### DISCUSSION TITLE
Question about `enabled: false`

### DISCUSSION THREAD

**Discussion by csantos-nydig** [TanStack/query#7855] [Q&A] [answered]
From https://github.com/TanStack/query/issues/1196

here is a quick and dirty example: https://codesandbox.io/p/sandbox/elegant-cdn-5hhvhh

```js
const useDataQuery = (props) => {
  return useSuspenseQuery({
    queryFn: () => {
      return { a: 1, b: 2, c: 3 };
    },
    queryKey: ["test"],
    ...props,
  });
};

const ComponentOne = () => {
  const { data } = useDataQuery();

  return <div>ComponentOne {JSON.stringify(data)} </div>;
};

const ComponentTwo = () => {
  const { data } = useDataQuery({
    enabled: false,
    select: (data) => {
      console.log("inside Component2 select method");
      return data.a + data.b;
    },
  });

  return <div>ComponentTwo {JSON.stringify(data)} </div>;
};
```

The query in ComponentTwo is forever `enabled: false`, yet the `select` method is executed and data is returned.

I'd expect ComponentTwo data to be undefined for ever until I change `enabled` to true; but it might be misunderstanding how it's supposed to work? 

----

Original response from @TkDodo https://github.com/TanStack/query/issues/7721#issuecomment-2266635167 (because I messed up and posted it in the wrong issue)

> But yes, you are misunderstanding how `enabled` works. When calling `useQuery`, you are always subscribing to the queryKey. If data changes, your component re-renders. `enabled` just makes sure that this query is "read-only", meaning it won't trigger any refetches by itself.

**Comment by csantos-nydig** [comment]
The follow up question I had (in the wrong thread, and why I decided to create this discussion instead)

is there any way to have the query in `ComponentTwo` to not execute the `select` at all if the query is disabled?

To me it's kind of suboptimal that the logic in `select` function is executed when the query is disabled

**Reply by csantos-nydig** [reply]
Sorry, it's just that the wording keeps it difficult to understand 😅 when I read `enabled: false` it makes me think as it like: "this is fully turned off until otherwise"; but based on your response it seems to have a different meaning.

So my question still exists: "Is it possible to have a query fully turned off until otherwise"?

**Reply by TkDodo** [reply]
no. you would to have your component `return null` if the query is not enabled. I'm wondering what the use-case for something like that would be ?

**Reply by csantos-nydig** [reply]
Sure, here is a pseudo code for a more "realistic" example:

```tsx
const useProductsQuery = (props) => {
  return useSuspenseQuery({
    queryFn: fetch('/api/products'),
    queryKey: ['products'],
    ...props
  })
}
```
```tsx
const ProductsList = () => {
  const { data = [] } = useProductsQuery();

  // list of products
  return <ul><li /></ul>
}
```
```tsx
const ProductDetails = ({ id }) => {
  const { data: productDetails } = useProductsQuery({
    enabled: !!id, // <----- this
    select: (products) => products.find(x => x.productId === id)
  })
}
```
```
<>
  <ProductList />
  <ProductDetails id="existing" />
  <ProductDetails id="not-existing" />
  <ProductDetails />
```

So I was expecting to save some time in `ProductDetails` to not execute the query (or at least not execute the `select` method) when the `id` isn't provided.

**Reply by csantos-nydig** [reply]
btw, I'm getting this expectation from the docs: https://tanstack.com/query/latest/docs/framework/react/guides/dependent-queries 
(I know this is from dependent queries, but it's the same concept imo)

<img width="479" alt="image" src="https://github.com/user-attachments/assets/31269f05-52f7-4dcc-805d-b19c7464cb4f">

**Reply by TkDodo** [reply, marked as answer]
> or at least not execute the select method

sorry, that's just not how it works. It also seems to be an "optimization" that isn't necessary. There is nothing to optimize here. Just remove the `enabled` and you're fine 🤷 

Btw: `enabled` doesn't exist on `useSuspenseQuery`.

**Reply by raix** [reply]
It seems like the documentation is still not updated - our team got hit by this question and it sparks confusion.

Maybe we could add an example of how to conditionally load data for a search or combobox e.g. Only triggering a query when search field was at least 3 characters seems like a common use case?

ref: [use enable or skipToken](https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#typesafe-disabling-of-queries-using-skiptoken)

Kind regards

Morten

**Reply by TkDodo** [reply]
that’s pretty much this example with a different `enabled` condition: https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#lazy-queries

**Comment by kristian-nst** [comment]
Same here. Just testing with `enabled: false` and  it still runs (originally `enabled: !!someDataExist`)

```
const {
  data: maxAmountData,
  isFetching: isMaxAmountFetching,
  error: maxAmountError,
  refetch: refetchMaxAmount,
} = apiClient.campaigns.maxAmount.useQuery({
  queryKey: [campaignsContract.maxAmount.path, slug],
  queryData: {
    query: {
      id: slug,
      walletId: walletData?.[0]?.id ?? '',
    },
  },
  enabled: false // example: !!walletData?.[0]?.id
})
```

**Reply by TkDodo** [reply]
same what? the original question was about running `select`, which your example doesn’t have ...

**Reply by kristian-nst** [reply]
Sorry, then I misunderstood :)
